### PR TITLE
cleanup block indices for missing blocks

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

For any existing nodes impacted by the block cleanup bug fixed [recently](https://github.com/prysmaticlabs/prysm/pull/15011), this is a hack to help those nodes self-heal and avoid a complete db wipe and resync.

**Notes for reviewers **

`blockParentRootIndicesBucket` maps from a block root to its children. When the db is in this state, we don't have the parent root of the missing block to clean up the index going the other direction. As I was wrapping this PR up I did think a couple ways to do it. 

1) Before we delete the slot -> root index, we can seek a db cursor to the key for the slot of the dangling block. We can then use Prev to look at the previous slot, which should contain the parent (there can be multiple block roots backed into the slot->root index). Check each of those keys and remove the deleted root if found.
2) Since we're looping over slot/root pairs in `blocksForSlotRange`, we could keep track of roots in the previous slot so that when we add something to the `badBlocks` slice for cleanup we have parent roots to check for the missing root. If the block root is in the first slot of the range we won't have seen its possible parents and need to fall back to method 1, so this is really just an optimization for method 1 and might not be worth the complexity.

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

### CODE REVIEW

File: beacon-chain/db/kv/blocks.go [hunk 1]
```diff
@@ -124,6 +125,10 @@ func (s *Store) Blocks(ctx context.Context, f *filters.QueryFilter) ([]interface
 	blocks := make([]interfaces.ReadOnlySignedBeaconBlock, 0)
 	blockRoots := make([][32]byte, 0)
 
+	if start, end, isSimple := f.SimpleSlotRange(); isSimple {
```
Comment (nisdas) [inline beacon-chain/db/kv/blocks.go:128]: Why is this added ? Doesn't the old filter work

File: beacon-chain/db/kv/blocks.go [hunk 2]
```diff
@@ -146,6 +151,69 @@ func (s *Store) Blocks(ctx context.Context, f *filters.QueryFilter) ([]interface
...
+// we don't know their parent root to efficiently clean the index going the other direction.
+func (s *Store) cleanupMissingBlockIndices(ctx context.Context, badBlocks []slotRoot) {
+	errs := make([]error, 0)
+	err := s.db.Update(func(tx *bolt.Tx) error {
+		for _, sr := range badBlocks {
+			log.WithField("root", fmt.Sprintf("%#x", sr.root)).WithField("slot", sr.slot).Warn("Cleaning up indices for missing block")
+			if err := s.deleteSlotIndexEntry(tx, sr.slot, sr.root); err != nil {
+				errs = append(errs, errors.Wrapf(err, "failed to clean up slot index entry for root %#x and slot %d", sr.root, sr.slot))
+			}
+			if err := tx.Bucket(blockParentRootIndicesBucket).Delete(sr.root[:]); err != nil {
```
Comment (nisdas) [inline beacon-chain/db/kv/blocks.go:165]: isn't this wrong ? For reference, how it was done in your earlier pr:
https://github.com/prysmaticlabs/prysm/pull/15011/files#diff-2396f91dd1a75091c28f42427f054081aae282cc7dd9ab9a7bb72fd32f9cec1dR1081

File: beacon-chain/db/kv/blocks.go [hunk 3]
```diff
@@ -729,10 +798,26 @@ func (s *Store) slotRootsInRange(ctx context.Context, start, end primitives.Slot
...
 		c := bkt.Cursor()
-		for k, v := c.Seek(key); k != nil; k, v = c.Prev() {
+		for k, v := c.Seek(key); ; /* rely on internal checks to exit */ k, v = c.Prev() {
+			if len(k) == 0 && len(v) == 0 {
+				// The `edge`` variable and this `if` deal with 2 edge cases:
+				// - Seeking past the end of the bucket (the `end` param is higher than the highest slot).
+				// - Seeking before the beginning of the bucket (the `start` param is lower than the lowest slot).
+				// In both of these cases k,v will be nil and we can handle the same way using `edge` to
+				// - continue to the next iteration. If the following Prev() key/value is nil, Prev has gone past the beginning.
+				// - Otherwise, iterate as usual.
```
Comment (nisdas) [inline beacon-chain/db/kv/blocks.go:813]: Do we have a unit test for these particular edge cases ?

Comment (kasey) [reply beacon-chain/db/kv/blocks.go:813]: Yes, `TestStore_Blocks_FiltersCorrectly` that touch the upper bounds of the blocks with `SetStartSlot` + `SetEndSlot`. LMK if that is sufficient, or if you think we need a separate test.

File: beacon-chain/db/kv/blocks.go (hunk 1 above)
Comment (kasey) [reply beacon-chain/db/kv/blocks.go:128]: The old filter only gives us the roots. To delete from the index, we also need the slot. We are obtaining the roots by going over the slot->root index, so we get the slot mapping for free by using the other method (which returns a slice of `slotRoot` structs).

File: beacon-chain/db/filters/filter.go [hunk 4]
```diff
@@ -132,3 +132,21 @@ func (q *QueryFilter) SetSlotStep(val uint64) *QueryFilter {
...
+	start, ok := q.queries[StartSlot].(primitives.Slot)
+	if !ok {
+		return 0, 0, false
+	}
+	end, ok := q.queries[EndSlot].(primitives.Slot)
+	if !ok {
+		return 0, 0, false
+	}
+	return start, end, true
+}
```
Comment (prestonvanloon) [inline beacon-chain/db/filters/filter.go:152]: Good candidate for unit test

File: beacon-chain/db/kv/blocks.go (hunk 2 above)
Comment (kasey) [reply beacon-chain/db/kv/blocks.go:165]: This index stores the mapping from a parent root to its children. So for a given root, there are 2 directions to consider:
- Removing the mapping *from* the block as parent to its children. That is what is happening on this line. Its straightforward because we know the root for the parent side of the relationship (the block to be deleted).
- Removing the mapping *to* the block as child, from its parent. The previous PR assumes we still have the block, which makes this straightforward because we can look up the block and then get its parent root. In this case where the block is missing, the block is deleted, so we don't know the parent without searching for it in the slot index. We can use the slot index to try to find it, but one complication there is that the parent may be a deep reorg on another fork - and maybe it was actually already deleted itself - so we don't quite know where to stop looking for it in the parent index. We could try scanning back a set amount of slots and give up. For now I left deleting the parent->block mapping as a TODO.

The previous PR diff doesn't include the "block as parent" deletion because that was already implemented in `DeleteBlock`.

**Review by nisdas — DISMISSED** [review]
LGTM

**Review by nisdas — DISMISSED** [review]

File: beacon-chain/db/filters/filter.go (hunk 4 above)
Comment (kasey) [reply beacon-chain/db/filters/filter.go:152]: added test

**Review by prestonvanloon — APPROVED** [review]
